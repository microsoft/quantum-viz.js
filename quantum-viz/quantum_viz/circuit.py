from contextlib import contextmanager
from typing import Any
from typing import Dict
from typing import List


class Circuit:
    """Python representation of a quantum circuit.
    Usage:
        circ = Circuit() # Create a circuit object
        circ.operation("H", [0]) # Add a gate named "H" that acts on a qubit [0]
        circ.to_json() # Generate JSON for usage with quantum-viz.js
    """

    def __init__(self, value: Dict[str, Any] = None) -> None:
        self._value = value
        self.qubits_to_bits = {}
        self.operations = []

    @staticmethod
    def _ids_to_dicts(ids: List[int], registers: List[int] = None):
        """Convert qubit IDs to a list of dicts"""
        if registers:
            return [{"type": 1, "qId": q, "cId": c} for q, c in zip(ids, registers)]
        return [{"qId": q} for q in ids]

    def create_qubits(self):
        def to_qubit_dict(qid, num):
            if num:
                return {"id": qid, "numChildren": num}
            return {"id": qid}

        return [
            to_qubit_dict(qid, num) for qid, num in sorted(self.qubits_to_bits.items())
        ]

    @property
    def qubits(self):
        return self.create_qubits()

    def to_json(self):
        return {"qubits": self.qubits, "operations": self.operations}

    def qubit(self, qubit_id: int, num_children: int = 0):
        if qubit_id not in self.qubits_to_bits:
            self.qubits_to_bits[qubit_id] = num_children
        if num_children:
            self.qubits_to_bits[qubit_id] = num_children

    def operation(self, name: str, targets: List[int]) -> None:
        self.controlled_op(name=name, controls=None, targets=targets)

    def measure(self, qubits: List[int], registers: List[int] = None):
        if registers is None:
            registers = [0] * len(qubits)
        self.controlled_op(
            name="Measure", controls=qubits, targets=qubits, registers=registers
        )

    @contextmanager
    def conditional_op(
        self,
        name: str,
        controls: List[int],
        registers: List[int] = None,
        conditional: int = 2,
    ):
        for qid in controls:
            self.qubit(qid)
        sub_circuit = Circuit()
        yield sub_circuit
        if registers is None:
            registers = [0] * len(controls)
        _controls = self._ids_to_dicts(controls, registers)
        [
            op.update({"conditionalRender": conditional, "controls": _controls})
            for op in sub_circuit.operations
        ]
        op = {
            "gate": name,
            "isConditional": "True",
            "targets": [],
            "controls": _controls,
            "children": sub_circuit.operations,
        }
        self.qubits_to_bits.update(sub_circuit.qubits_to_bits)
        self.operations.append(op)

    def controlled_op(
        self,
        name: str,
        controls: List[int],
        targets: List[int],
        registers: List[int] = None,
    ) -> None:
        op = {"gate": name, "targets": self._ids_to_dicts(targets, registers)}
        if "Measure" in name:
            op.update(
                {"isMeasurement": "True", "controls": self._ids_to_dicts(targets)}
            )
            for qid in targets:
                self.qubit(qid, 1)
        elif controls:
            if controls is not None:
                op.update(
                    {"isControlled": "True", "controls": self._ids_to_dicts(controls)}
                )
                for qid in targets + controls:
                    self.qubit(qid)
            else:
                raise ValueError("No controls specified.")
        else:
            for qid in targets:
                self.qubit(qid)
        self.operations.append(op)

    def add_group(self, name: str, sub_circuit: "Circuit", qubits: List[int] = None):
        qubits: Dict[int, int] = sub_circuit.qubits_to_bits
        operations: List[Dict[str, Any]] = sub_circuit.operations
        op = {
            "gate": name,
            "children": operations,
            "targets": self._ids_to_dicts(qubits.keys()),
        }
        self.qubits_to_bits.update(qubits)
        self.operations.append(op)

    @contextmanager
    def group(self, name: str):
        sub_circuit = Circuit()
        yield sub_circuit
        self.add_group(name, sub_circuit)

    def _ipython_display_(self) -> None:
        from quantum_viz.widget import QViz

        return QViz(self.to_json())._ipython_display_()
