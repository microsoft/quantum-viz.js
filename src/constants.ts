// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

// SVG Namespace
export const svgNS = 'http://www.w3.org/2000/svg';

// Display attributes
/** Left padding of SVG. */
export const leftPadding = 20;
/** x coordinate for first operation on each register. */
export const startX = 80;
/** y coordinate of first register. */
export const startY = 40;
/** Minimum width of each gate. */
export const minGateWidth = 40;
/** Height of each gate. */
export const gateHeight = 40;
/** Padding on each side of gate. */
export const gatePadding = 10;
/** Padding on each side of gate label. */
export const labelPadding = 10;
/** Height between each qubit register. */
export const registerHeight: number = gateHeight + gatePadding * 2;
/** Height between classical registers. */
export const classicalRegHeight: number = gateHeight;
/** Group box inner padding. */
export const groupBoxPadding = gatePadding;
/** Padding between nested groups. */
export const nestedGroupPadding = 2;
/** Additional offset for control button. */
export const controlBtnOffset = 40;
/** Control button radius. */
export const controlBtnRadius = 15;
/** Default font size for gate labels. */
export const labelFontSize = 14;
/** Default font size for gate arguments. */
export const argsFontSize = 12;
/** Starting x coord for each register wire. */
export const regLineStart = 40;

// Panel
/** Panel width */
export const panelWidth = 144;
/** Gap between gates in Add Panel */
export const gateGap = 50;
