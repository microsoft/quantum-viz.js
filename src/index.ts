// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { Sqore } from './sqore';
import { Circuit } from './circuit';
import { StyleConfig } from './styles';

/**
 * Render `circuit` into `container` at the specified layer depth.
 *
 * @param circuit Circuit to be visualized.
 * @param container HTML element for rendering visualization into.
 * @param style Custom visualization style.
 * @param renderDepth Initial layer depth at which to render gates.
 * @param isEditable Optional value enabling/disabling editable feature
 * @param onCircuitChange Optional function to trigger when changing elements in circuit
 */
export const draw = (
    circuit: Circuit,
    container: HTMLElement,
    style: StyleConfig | string = {},
    renderDepth = 0,
    isEditable?: boolean,
    onCircuitChange?: (circuit: Circuit) => void,
): void => {
    const sqore = new Sqore(circuit, style);

    sqore.draw(container, renderDepth, isEditable, onCircuitChange);
};

export { STYLES } from './styles';

// Export types
export type { StyleConfig } from './styles';
export type { Circuit, Qubit, Operation } from './circuit';
