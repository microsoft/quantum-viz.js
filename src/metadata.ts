// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { DataAttributes } from './circuit';

/**
 * Enum for the various gate operations handled.
 */
export enum GateType {
    /** Measurement gate. */
    Measure,
    /** CNOT gate. */
    Cnot,
    /** SWAP gate. */
    Swap,
    /** Single/multi qubit unitary gate. */
    Unitary,
    /** Single/multi controlled unitary gate. */
    ControlledUnitary,
    /** Nested group of classically-controlled gates. */
    ClassicalControlled,
    /** Group of nested gates */
    Group,
    /** Invalid gate. */
    Invalid,
}

/**
 * Metadata used to store information pertaining to a given
 * operation for rendering its corresponding SVG.
 */
export interface Metadata {
    /** Gate type. */
    type: GateType;
    /** Centre x coord for gate position. */
    x: number;
    /** Array of y coords of control registers. */
    controlsY: number[];
    /** Array of y coords of target registers.
     *  For `GateType.Unitary` or `GateType.ControlledUnitary`, this is an array of groups of
     *  y coords, where each group represents a unitary box to be rendered separately.
     */
    targetsY: (number | number[])[];
    /** Gate label. */
    label: string;
    /** Gate arguments as string. */
    displayArgs?: string;
    /** Gate width. */
    width: number;
    /** Children operations as part of group. */
    children?: (Metadata | Metadata[])[];
    /** Custom data attributes to attach to gate element. */
    dataAttributes?: DataAttributes;
}
