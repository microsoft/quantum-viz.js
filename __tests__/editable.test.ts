import { Operation } from '../src/circuit';
import { exportedForTesting } from '../src/editable';
import { RegisterType } from '../src/register';

const {
    getDataId,
    splitDataId,
    cursorMove,
    cursorCopy,
    deleteAt,
    insertBefore,
    insertAfter,
    getDropzonePosition,
    getWireElemText,
    getWireElemY,
    getGate,
    getParent,
} = exportedForTesting;

// Utlities
describe('Testing getDataId', () => {
    const elem = document.createElement('div');
    test('with with no data-id', () => {
        expect(getDataId(elem)).toBe('');
    });
    test('with level 0 data-id', () => {
        elem.setAttribute('data-id', '0');
        expect(getDataId(elem)).toBe('0');
    });

    test('with level 1 data-id', () => {
        elem.setAttribute('data-id', '0-1');
        expect(getDataId(elem)).toBe('0-1');
    });
});

describe('Testing splitDataId', () => {
    test('with empty dataId', () => {
        expect(splitDataId('')).toStrictEqual([]);
    });
    test('with level 0 data-id', () => {
        expect(splitDataId('1')).toStrictEqual([1]);
    });

    test('with level 1 data-id', () => {
        expect(splitDataId('1-2')).toStrictEqual([1, 2]);
    });
});

describe('Testing cursorMove', () => {
    const container = document.createElement('div');
    test('turn on move cursor', () => {
        expect(container).toMatchSnapshot();
        cursorMove(container, true);
        expect(container).toMatchSnapshot();
    });
    test('turn off move cursor', () => {
        expect(container).toMatchSnapshot();
        cursorMove(container, false);
        expect(container).toMatchSnapshot();
    });
    test('turn on and off move cursor', () => {
        expect(container).toMatchSnapshot();
        cursorMove(container, true);
        cursorMove(container, false);
        expect(container).toMatchSnapshot();
    });
});

describe('Testing cursorCopy', () => {
    const container = document.createElement('div');
    test('turn on copy cursor', () => {
        expect(container).toMatchSnapshot();
        cursorCopy(container, true);
        expect(container).toMatchSnapshot();
    });
    test('turn off copy cursor', () => {
        expect(container).toMatchSnapshot();
        cursorCopy(container, false);
        expect(container).toMatchSnapshot();
    });
    test('turn on and off copy cursor', () => {
        expect(container).toMatchSnapshot();
        cursorCopy(container, true);
        cursorCopy(container, false);
        expect(container).toMatchSnapshot();
    });
});

describe('Testing deleteAt', () => {
    const operations: Operation[] = [];
    beforeEach(() => {
        Object.assign(operations, [
            {
                gate: 'X',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
            {
                gate: 'Y',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
            {
                gate: 'Z',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
        ]);
    });
    test('delete X at index 0', () => {
        deleteAt(operations, 0);
        expect(operations).toStrictEqual([
            {
                gate: 'Y',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
            {
                gate: 'Z',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
        ]);
    });
    test('delete Y at index 1', () => {
        deleteAt(operations, 1);
        expect(operations).toStrictEqual([
            {
                gate: 'X',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
            {
                gate: 'Z',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
        ]);
    });
    test('delete Z and X at index 2, 0', () => {
        deleteAt(operations, 2);
        deleteAt(operations, 0);
        expect(operations).toStrictEqual([
            {
                gate: 'Y',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
        ]);
    });
});

describe('Testing insertBefore', () => {
    test('insert before X', () => {
        const operations = [
            {
                gate: 'X',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
            {
                gate: 'Z',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
        ];
        const newGate = {
            gate: 'Y',
            isMeasurement: false,
            isConditional: false,
            isControlled: false,
            isAdjoint: false,
            controls: [],
            targets: [{ type: RegisterType.Qubit, qId: 0 }],
        };
        insertBefore(operations, 0, newGate);
        expect(operations).toStrictEqual([
            {
                gate: 'Y',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
            {
                gate: 'X',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
            {
                gate: 'Z',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
        ]);
    });
});

describe('Testing insertAfter', () => {
    test('insert after X', () => {
        const operations = [
            {
                gate: 'X',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
            {
                gate: 'Z',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
        ];
        const newGate = {
            gate: 'Y',
            isMeasurement: false,
            isConditional: false,
            isControlled: false,
            isAdjoint: false,
            controls: [],
            targets: [{ type: RegisterType.Qubit, qId: 0 }],
        };
        insertAfter(operations, 0, newGate);
        expect(operations).toStrictEqual([
            {
                gate: 'X',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
            {
                gate: 'Y',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
            {
                gate: 'Z',
                isMeasurement: false,
                isConditional: false,
                isControlled: false,
                isAdjoint: false,
                controls: [],
                targets: [{ type: RegisterType.Qubit, qId: 0 }],
            },
        ]);
    });
});

describe('Testing getDropzonePosition', () => {
    let svgElem: SVGElement;
    let dropzoneElem: SVGElement;
    beforeEach(() => {
        svgElem = document.createElementNS('http://www.w3.org/2000/svg', 'svg') as SVGElement;
        dropzoneElem = document.createElementNS('http://www.w3.org/2000/svg', 'rect') as SVGElement;
        svgElem.append(dropzoneElem);
    });
    test('get position of non-dropzone', () => {
        expect(() => getDropzonePosition(dropzoneElem)).toThrowError('Position not found');
    });
    test('get position of dropzone on the left', () => {
        dropzoneElem.setAttribute('data-dropzone-position', 'left');
        expect(getDropzonePosition(dropzoneElem)).toBe('left');
    });
    test('get position of dropzone on the right', () => {
        dropzoneElem.setAttribute('data-dropzone-position', 'right');
        expect(getDropzonePosition(dropzoneElem)).toBe('right');
    });
});

describe('Testing getWireElementText', () => {
    let svgElem: SVGElement;
    let groupElem: SVGGElement;
    let textElem: SVGGElement;
    beforeEach(() => {
        svgElem = document.createElementNS('http://www.w3.org/2000/svg', 'svg') as SVGElement;
        groupElem = document.createElementNS('http://www.w3.org/2000/svg', 'g') as SVGGElement;
        textElem = document.createElementNS('http://www.w3.org/2000/svg', 'text') as SVGTextElement;
        groupElem.append(textElem);
        svgElem.append(groupElem);
    });
    test('text element not exists', () => {
        textElem.remove();
        expect(() => getWireElemText(groupElem)).toThrowError('Text not found');
    });
    test('get text element without textContent', () => {
        expect(() => getWireElemText(groupElem)).toThrowError('Text not found');
    });
    test('get text element empty textContent', () => {
        expect(() => getWireElemText(groupElem)).toThrowError('Text not found');
    });
    test('should return q0', () => {
        textElem.textContent = 'q0';
        expect(getWireElemText(groupElem)).toEqual('q0');
    });
    test('should return q1', () => {
        textElem.textContent = 'q1';
        expect(getWireElemText(groupElem)).toEqual('q1');
    });
});

describe('Testing getWireElemY', () => {
    let svgElem: SVGElement;
    let groupElem: SVGGElement;
    let lineElem: SVGGElement;
    beforeEach(() => {
        svgElem = document.createElementNS('http://www.w3.org/2000/svg', 'svg') as SVGElement;
        groupElem = document.createElementNS('http://www.w3.org/2000/svg', 'g') as SVGGElement;
        lineElem = document.createElementNS('http://www.w3.org/2000/svg', 'line') as SVGLineElement;
        groupElem.append(lineElem);
        svgElem.append(groupElem);
    });
    test('line element not exists', () => {
        lineElem.remove();
        expect(() => getWireElemY(groupElem)).toThrowError('y not found');
    });
    test('get y element without y value', () => {
        expect(() => getWireElemY(groupElem)).toThrowError('y not found');
    });
    test('get text element empty textContent', () => {
        expect(() => getWireElemY(groupElem)).toThrowError('y not found');
    });
    test('should return 40', () => {
        lineElem.setAttribute('y1', '40');
        expect(getWireElemY(groupElem)).toEqual(40);
    });
    test('should return 99', () => {
        lineElem.setAttribute('y1', '99');
        expect(getWireElemY(groupElem)).toEqual(99);
    });
});

describe('Testing getParent', () => {
    test('with level 0 gate', () => {
        const operations = [
            {
                gate: 'H',
                targets: [{ qId: 0 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 0 }],
                targets: [{ qId: 1 }],
            },
            {
                gate: 'Measure',
                isMeasurement: true,
                controls: [{ qId: 1 }],
                targets: [{ type: 1, qId: 1, cId: 0 }],
            },
        ];
        expect(getParent('0', operations)).toStrictEqual([
            {
                gate: 'H',
                targets: [{ qId: 0 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 0 }],
                targets: [{ qId: 1 }],
            },
            {
                gate: 'Measure',
                isMeasurement: true,
                controls: [{ qId: 1 }],
                targets: [{ type: 1, qId: 1, cId: 0 }],
            },
        ]);
    });
    test('with level 1 gate', () => {
        const operations = [
            {
                gate: 'Foo',
                conditionalRender: 3,
                targets: [{ qId: 0 }, { qId: 1 }],
                children: [
                    {
                        gate: 'H',
                        targets: [{ qId: 1 }],
                    },
                    {
                        gate: 'RX',
                        displayArgs: '(0.25)',
                        isControlled: true,
                        controls: [{ qId: 1 }],
                        targets: [{ qId: 0 }],
                    },
                ],
            },
            {
                gate: 'X',
                targets: [{ qId: 3 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 1 }],
                targets: [{ qId: 2 }, { qId: 3 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 2 }, { qId: 3 }],
                targets: [{ qId: 1 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 1 }, { qId: 3 }],
                targets: [{ qId: 2 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 2 }],
                targets: [{ qId: 1 }, { qId: 3 }],
            },
            {
                gate: 'measure',
                isMeasurement: true,
                controls: [{ qId: 0 }],
                targets: [{ type: 1, qId: 0, cId: 0 }],
            },
            {
                gate: 'ApplyIfElseR',
                isConditional: true,
                controls: [{ type: 1, qId: 0, cId: 0 }],
                targets: [],
                children: [
                    {
                        gate: 'H',
                        targets: [{ qId: 1 }],
                        conditionalRender: 1,
                    },
                    {
                        gate: 'X',
                        targets: [{ qId: 1 }],
                        conditionalRender: 1,
                    },
                    {
                        gate: 'X',
                        isControlled: true,
                        controls: [{ qId: 0 }],
                        targets: [{ qId: 1 }],
                        conditionalRender: 2,
                    },
                    {
                        gate: 'Foo',
                        targets: [{ qId: 3 }],
                        conditionalRender: 2,
                    },
                ],
            },
            {
                gate: 'SWAP',
                targets: [{ qId: 0 }, { qId: 2 }],
                children: [
                    { gate: 'X', isControlled: true, controls: [{ qId: 0 }], targets: [{ qId: 2 }] },
                    { gate: 'X', isControlled: true, controls: [{ qId: 2 }], targets: [{ qId: 0 }] },
                    { gate: 'X', isControlled: true, controls: [{ qId: 0 }], targets: [{ qId: 2 }] },
                ],
            },
            {
                gate: 'ZZ',
                targets: [{ qId: 1 }, { qId: 3 }],
            },
            {
                gate: 'ZZ',
                targets: [{ qId: 0 }, { qId: 1 }],
            },
            {
                gate: 'XX',
                isControlled: true,
                controls: [{ qId: 0 }],
                targets: [{ qId: 1 }, { qId: 3 }],
            },
            {
                gate: 'XX',
                isControlled: true,
                controls: [{ qId: 2 }],
                targets: [{ qId: 1 }, { qId: 3 }],
            },
            {
                gate: 'XX',
                isControlled: true,
                controls: [{ qId: 0 }, { qId: 2 }],
                targets: [{ qId: 1 }, { qId: 3 }],
            },
        ];
        expect(getParent('0-1', operations)).toStrictEqual([
            {
                gate: 'H',
                targets: [{ qId: 1 }],
            },
            {
                gate: 'RX',
                displayArgs: '(0.25)',
                isControlled: true,
                controls: [{ qId: 1 }],
                targets: [{ qId: 0 }],
            },
        ]);
    });
});

describe('Testing getGate', () => {
    test('should return H gate', () => {
        const operations = [
            {
                gate: 'H',
                targets: [{ qId: 0 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 0 }],
                targets: [{ qId: 1 }],
            },
            {
                gate: 'Measure',
                isMeasurement: true,
                controls: [{ qId: 1 }],
                targets: [{ type: 1, qId: 1, cId: 0 }],
            },
        ];
        expect(getGate('0', operations)).toStrictEqual({
            gate: 'H',
            targets: [{ qId: 0 }],
        });
    });
    test('should return X gate', () => {
        const operations = [
            {
                gate: 'H',
                targets: [{ qId: 0 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 0 }],
                targets: [{ qId: 1 }],
            },
            {
                gate: 'Measure',
                isMeasurement: true,
                controls: [{ qId: 1 }],
                targets: [{ type: 1, qId: 1, cId: 0 }],
            },
        ];
        expect(getGate('1', operations)).toStrictEqual({
            gate: 'X',
            isControlled: true,
            controls: [{ qId: 0 }],
            targets: [{ qId: 1 }],
        });
    });
    test('should return RX', () => {
        const operations = [
            {
                gate: 'Foo',
                conditionalRender: 3,
                targets: [{ qId: 0 }, { qId: 1 }],
                children: [
                    {
                        gate: 'H',
                        targets: [{ qId: 1 }],
                    },
                    {
                        gate: 'RX',
                        displayArgs: '(0.25)',
                        isControlled: true,
                        controls: [{ qId: 1 }],
                        targets: [{ qId: 0 }],
                    },
                ],
            },
            {
                gate: 'X',
                targets: [{ qId: 3 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 1 }],
                targets: [{ qId: 2 }, { qId: 3 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 2 }, { qId: 3 }],
                targets: [{ qId: 1 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 1 }, { qId: 3 }],
                targets: [{ qId: 2 }],
            },
            {
                gate: 'X',
                isControlled: true,
                controls: [{ qId: 2 }],
                targets: [{ qId: 1 }, { qId: 3 }],
            },
            {
                gate: 'measure',
                isMeasurement: true,
                controls: [{ qId: 0 }],
                targets: [{ type: 1, qId: 0, cId: 0 }],
            },
            {
                gate: 'ApplyIfElseR',
                isConditional: true,
                controls: [{ type: 1, qId: 0, cId: 0 }],
                targets: [],
                children: [
                    {
                        gate: 'H',
                        targets: [{ qId: 1 }],
                        conditionalRender: 1,
                    },
                    {
                        gate: 'X',
                        targets: [{ qId: 1 }],
                        conditionalRender: 1,
                    },
                    {
                        gate: 'X',
                        isControlled: true,
                        controls: [{ qId: 0 }],
                        targets: [{ qId: 1 }],
                        conditionalRender: 2,
                    },
                    {
                        gate: 'Foo',
                        targets: [{ qId: 3 }],
                        conditionalRender: 2,
                    },
                ],
            },
            {
                gate: 'SWAP',
                targets: [{ qId: 0 }, { qId: 2 }],
                children: [
                    { gate: 'X', isControlled: true, controls: [{ qId: 0 }], targets: [{ qId: 2 }] },
                    { gate: 'X', isControlled: true, controls: [{ qId: 2 }], targets: [{ qId: 0 }] },
                    { gate: 'X', isControlled: true, controls: [{ qId: 0 }], targets: [{ qId: 2 }] },
                ],
            },
            {
                gate: 'ZZ',
                targets: [{ qId: 1 }, { qId: 3 }],
            },
            {
                gate: 'ZZ',
                targets: [{ qId: 0 }, { qId: 1 }],
            },
            {
                gate: 'XX',
                isControlled: true,
                controls: [{ qId: 0 }],
                targets: [{ qId: 1 }, { qId: 3 }],
            },
            {
                gate: 'XX',
                isControlled: true,
                controls: [{ qId: 2 }],
                targets: [{ qId: 1 }, { qId: 3 }],
            },
            {
                gate: 'XX',
                isControlled: true,
                controls: [{ qId: 0 }, { qId: 2 }],
                targets: [{ qId: 1 }, { qId: 3 }],
            },
        ];
        expect(getGate('0-1', operations)).toStrictEqual({
            gate: 'RX',
            displayArgs: '(0.25)',
            isControlled: true,
            controls: [{ qId: 1 }],
            targets: [{ qId: 0 }],
        });
    });
});
