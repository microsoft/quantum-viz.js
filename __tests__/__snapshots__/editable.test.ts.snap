// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test _copyX copy elem from index 0 to index 1 1`] = `
Array [
  Object {
    "gate": "H",
    "targets": Array [
      Object {
        "qId": 0,
      },
    ],
  },
  Object {
    "controls": Array [
      Object {
        "qId": 0,
      },
    ],
    "gate": "X",
    "isControlled": true,
    "targets": Array [
      Object {
        "qId": 1,
      },
    ],
  },
  Object {
    "gate": "H",
    "targets": Array [
      Object {
        "qId": 0,
      },
    ],
  },
  Object {
    "controls": Array [
      Object {
        "qId": 1,
      },
    ],
    "gate": "Measure",
    "isMeasurement": true,
    "targets": Array [
      Object {
        "cId": 0,
        "qId": 1,
        "type": 1,
      },
    ],
  },
]
`;

exports[`Test _copyX copy elem from index 0 to last 1`] = `
Array [
  Object {
    "gate": "H",
    "targets": Array [
      Object {
        "qId": 0,
      },
    ],
  },
  Object {
    "controls": Array [
      Object {
        "qId": 0,
      },
    ],
    "gate": "X",
    "isControlled": true,
    "targets": Array [
      Object {
        "qId": 1,
      },
    ],
  },
  Object {
    "controls": Array [
      Object {
        "qId": 1,
      },
    ],
    "gate": "Measure",
    "isMeasurement": true,
    "targets": Array [
      Object {
        "cId": 0,
        "qId": 1,
        "type": 1,
      },
    ],
  },
  Object {
    "gate": "H",
    "targets": Array [
      Object {
        "qId": 0,
      },
    ],
  },
]
`;

exports[`Test _equivGateElem should return gate H 1`] = `
<g
  class="gate"
  data-id="0"
  data-zoom-in="false"
  data-zoom-out="false"
>
  <g>
    <g>
      <rect
        class="gate-unitary"
        height="40"
        width="40"
        x="80"
        y="20"
      />
      <text
        font-size="14"
        x="100"
        y="40"
      >
        H
      </text>
    </g>
  </g>
</g>
`;

exports[`Test _equivOperation should return H gate 1`] = `
Object {
  "gate": "H",
  "targets": Array [
    Object {
      "qId": 0,
    },
  ],
}
`;

exports[`Test _equivOperation should return X gate 1`] = `
Object {
  "controls": Array [
    Object {
      "qId": 0,
    },
  ],
  "gate": "X",
  "isControlled": true,
  "targets": Array [
    Object {
      "qId": 1,
    },
  ],
}
`;

exports[`Test _equivOperationParent should return Foo 1`] = `
Array [
  Object {
    "gate": "H",
    "targets": Array [
      Object {
        "qId": 1,
      },
    ],
  },
  Object {
    "controls": Array [
      Object {
        "qId": 1,
      },
    ],
    "displayArgs": "(0.25)",
    "gate": "RX",
    "isControlled": true,
    "targets": Array [
      Object {
        "qId": 0,
      },
    ],
  },
]
`;

exports[`Test _equivOperationParent should return all operations 1`] = `
Array [
  Object {
    "gate": "H",
    "targets": Array [
      Object {
        "qId": 0,
      },
    ],
  },
  Object {
    "controls": Array [
      Object {
        "qId": 0,
      },
    ],
    "gate": "X",
    "isControlled": true,
    "targets": Array [
      Object {
        "qId": 1,
      },
    ],
  },
  Object {
    "controls": Array [
      Object {
        "qId": 1,
      },
    ],
    "gate": "Measure",
    "isMeasurement": true,
    "targets": Array [
      Object {
        "cId": 0,
        "qId": 1,
        "type": 1,
      },
    ],
  },
]
`;

exports[`Test _moveX move elem from index 0 to index 1 1`] = `
Array [
  Object {
    "controls": Array [
      Object {
        "qId": 0,
      },
    ],
    "gate": "X",
    "isControlled": true,
    "targets": Array [
      Object {
        "qId": 1,
      },
    ],
  },
  Object {
    "gate": "H",
    "targets": Array [
      Object {
        "qId": 0,
      },
    ],
  },
  Object {
    "controls": Array [
      Object {
        "qId": 1,
      },
    ],
    "gate": "Measure",
    "isMeasurement": true,
    "targets": Array [
      Object {
        "cId": 0,
        "qId": 1,
        "type": 1,
      },
    ],
  },
]
`;

exports[`Test _moveX move elem from index 0 to last 1`] = `
Array [
  Object {
    "controls": Array [
      Object {
        "qId": 0,
      },
    ],
    "gate": "X",
    "isControlled": true,
    "targets": Array [
      Object {
        "qId": 1,
      },
    ],
  },
  Object {
    "controls": Array [
      Object {
        "qId": 1,
      },
    ],
    "gate": "Measure",
    "isMeasurement": true,
    "targets": Array [
      Object {
        "cId": 0,
        "qId": 1,
        "type": 1,
      },
    ],
  },
  Object {
    "gate": "H",
    "targets": Array [
      Object {
        "qId": 0,
      },
    ],
  },
]
`;
